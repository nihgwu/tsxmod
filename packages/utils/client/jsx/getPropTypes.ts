import type { PropertySignature } from 'ts-morph'
import { Node, TypeFormatFlags } from 'ts-morph'
import { getDefaultValuesFromProperties } from '../properties'

/** Gets the prop types for a component declaration. */
export function getPropTypes(declaration: Node) {
  const signatures = declaration.getType().getCallSignatures()

  if (signatures.length === 0) {
    return null
  }

  const [propsSignature] = signatures
  const [props] = propsSignature.getParameters()

  if (props) {
    const valueDeclaration = props.getValueDeclaration()

    if (!valueDeclaration) {
      return null
    }

    const propsType = declaration
      .getProject()
      .getTypeChecker()
      .getTypeOfSymbolAtLocation(props, valueDeclaration)
    const firstChild = valueDeclaration.getFirstChild()
    let defaultValues: ReturnType<typeof getDefaultValuesFromProperties> = {}

    if (Node.isObjectBindingPattern(firstChild)) {
      defaultValues = getDefaultValuesFromProperties(firstChild.getElements())
    }

    return propsType
      .getApparentProperties()
      .map((prop) => {
        const declarations = prop.getDeclarations()
        const propDeclaration = declarations[0] as PropertySignature

        if (
          (propDeclaration || declaration)
            .getSourceFile()
            .getFilePath()
            .includes('node_modules')
        ) {
          return null
        }

        const propName = prop.getName()
        const propType = prop.getTypeAtLocation(declaration)
        const description = prop
          .getDeclarations()
          .filter(Node.isJSDocable)
          .map((declaration) =>
            declaration
              .getJsDocs()
              .map((doc) => doc.getComment())
              .flat()
          )
          .join('\n')
        const defaultValue = defaultValues[propName]

        return {
          name: propName,
          required: !propDeclaration?.hasQuestionToken() && !defaultValue,
          description: description || null,
          type: propType.getText(
            declaration,
            TypeFormatFlags.UseAliasDefinedOutsideCurrentScope
          ),
          defaultValue,
        }
      })
      .filter(Boolean)
  }

  return null
}
