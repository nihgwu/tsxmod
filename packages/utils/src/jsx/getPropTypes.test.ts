import { Project } from 'ts-morph'
import { getPropTypes } from './getPropTypes'

test('returns prop types for a React element', () => {
  const project = new Project({ useInMemoryFileSystem: true })
  const sourceFile = project.createSourceFile(
    'test.tsx',
    `
    import React from 'react'

    type Props = {
      foo: string
    }

    export const Component = (props: Props) => {
      return <div />
    }
    `
  )
  const component = sourceFile.getVariableDeclarationOrThrow('Component')
  const propTypes = getPropTypes(component)

  expect(propTypes).toMatchSnapshot()
})
