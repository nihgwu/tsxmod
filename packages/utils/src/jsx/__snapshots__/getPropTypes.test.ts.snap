// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns prop types for a React element 1`] = `
[
  {
    "defaultValue": null,
    "description": null,
    "name": "foo",
    "required": true,
    "type": "string",
  },
]
`;
